File: widgets/CorticalSurface/CorticalSurface.js
import React, { Component } from 'react';
import { PythonInterface } from 'reactopya';
import Surface3d from '../Surface3d/Surface3d';
const config = require('./CorticalSurface.json');

export default class CorticalSurface extends Component {
    static title = 'Render cortical surface in 3D from .nwb file'
    static reactopyaConfig = config
    constructor(props) {
        super(props);
        this.state = {
            // javascript state
            path: null,
            download_from: null,
            name: null,

            // python state (returned from python backend)
            status: '',
            status_message: '',
            vertices: null,
            faces: null,

            // see below
            surface: null
        }
    }
    componentDidMount() {
        this.pythonInterface = new PythonInterface(this, config);
        this.pythonInterface.setState({
            path: this.props.path,
            download_from: this.props.download_from,
            name: this.props.name
        });
        this.pythonInterface.start();
    }
    componentWillUnmount() {
        this.pythonInterface.stop();
    }
    render() {
        return (
            <RespectStatus {...this.state}>
                <Surface3d
                    {...this.props}
                    vertices={this.state.vertices}
                    faces={this.state.faces}
                />
            </RespectStatus>
        )
    }
}

class RespectStatus extends Component {
    state = {}
    render() {
        switch (this.props.status) {
            case 'running':
                return <div>Running: {this.props.status_message}</div>
            case 'error':
                return <div>Error: {this.props.status_message}</div>
            case 'finished':
                return this.props.children;
            default:
                return <div>Unknown status: {this.props.status}</div>
        }
    }
}

File: widgets/CorticalSurface/CorticalSurface.py
from vtk import vtkXMLPolyDataReader
from ..pycommon.nwb_to_dict import nwb_to_dict
from mountaintools import client as mt
import numpy as np
from .examples import examples


class CorticalSurface:
    examples = examples

    def __init__(self):
        super().__init__()

    def javascript_state_changed(self, prev_state, state):
        self.set_python_state(dict(status='running', status_message='Running'))

        # get javascript state
        download_from = state.get('download_from', [])
        path = state.get('path', None)
        name = state.get('name', None)

        if path and name:
            mt.configDownloadFrom(download_from)
            if path.endswith('.nwb'):
                self.set_python_state(dict(status_message='Realizing object from nwb file: {}'.format(path)))
                obj = nwb_to_dict(path, use_cache=True)
            else:
                self.set_python_state(dict(status_message='Realizing object: {}'.format(path)))
                obj = mt.loadObject(path=path)
            if not obj:
                self.set_python_state(dict(
                    status='error',
                    status_message='Unable to realize object: {}'.format(path)
                ))
                return
            datasets = obj['general']['subject']['cortical_surfaces'][name]['_datasets']
            faces0 = np.load(mt.realizeFile(datasets['faces']['_data']))
            vertices = np.load(mt.realizeFile(datasets['vertices']['_data'])).T

            # there's a better way to do the following
            # (need to get it into a single vector format)
            faces = []
            for j in range(faces0.shape[0]):
                # 3 = #vertices in polygon (assume a triangulation)
                faces.extend([3, faces0[j, 0], faces0[j, 1], faces0[j, 2]])
            faces = np.array(faces)

            # return this python state to the javascript
            self.set_python_state(dict(
                faces=faces,
                vertices=vertices,
                status='finished',
                status_message='Done.'
            ))
        else:
            self.set_python_state(dict(
                status='error',
                status_message='Missing path and/or name'
            ))
